### PR TITLE
Store resource in tracer SDK

### PR DESCRIPTION
**Pull request by pavolloffay**
Updates #281 

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java [hunk 1]
```diff
@@ -51,11 +53,13 @@ public Scope withSpan(Span span) {
   }
 
   @Override
-  public void setResource(Resource resource) {}
+  public void setResource(Resource resource) {
+    this.resource = resource;
+  }
 
   @Override
   public Resource getResource() {
```
Comment (SergeyKanzhelev) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: is there a comment on API to not cache this value? We need to clearly state that the object returned by this API can be replaced in runtime at any moment

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: Some questions:
1. I am not sure why do we need a get on this?
1. Also not sure we need a set, I think we can just offer the ability to set the initial resource during initialization time of the TracerSdk.

Comment (pavolloffay) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: I did this PR to move forward with the impl. Setter and getter on the resource seem not required in my opinion. I thought that this has been discussed as it is in the API.


I think that setting a resource belongs to initialization. Getter seems also redundant.

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: @pavolloffay I am sorry that I kind of moved the discussion a bit here and that it happen in your PR. You did a good job on this PR, but want to make sure we do the best of the project.

I think we are on the same opinion about what is needed here.

Comment (pavolloffay) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: So do we remove these methods completely fro the tracer API?

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: That's what I am proposing, waiting for @SergeyKanzhelev and @carlosalberto to see what do they think?

Comment (carlosalberto) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: Oh, I remember we talking about this some time ago ;) Yeah, it makes sense to have this as part of the `Tracer` initialization only - leaving the `get()` part would be fine, but probably we don't **really** need to expose it in the main API.

Comment (SergeyKanzhelev) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: One reason to expose it is to allow to inject this resource information in outgoing requests and response headers. There are systems that exchange this information.

Comment (carlosalberto) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TracerSdk.java]: Yeah, I'd propose go with the 'safe' route now and remove it for the moment, and add it if/when actually needed.

**Comment by bogdandrutu** [comment]
@SergeyKanzhelev are you approving to remove these or the current PR?

**Comment by SergeyKanzhelev** [comment]
oh, I didn't realize get for resource already was there. I glanced over and though that getter was removed so we only have `set`.

**Comment by bogdandrutu** [comment]
So, we agreed to remove the getter from Tracer. What about set, do we need an explicit set or we want only the SDK to allow that (not the API)?

I see no reason for an instrumentation plugin to set the resources globally, at most we need what we discussed in the components issue, to add more data about the instrumented component and merge these into the default tracer Resource.

**Comment by carlosalberto** [comment]
> So, we agreed to remove the getter from Tracer. What about set, do we need an explicit set or we want only the SDK to allow that (not the API)?

So I *feel* like we should set this only a ctor time (at least for the time being).

**Comment by SergeyKanzhelev** [comment]
> So I feel like we should set this only a ctor time (at least for the time being).

If Resoruce information is obtained from metadata REST call in [AWS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) and [Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service), would you require for the process to hold on on doing anything before that call will complete? Setter would allow to update Resource information later in the process.

**Comment by bogdandrutu** [comment]
@SergeyKanzhelev I think this is very specific to every implementation. I agree with you that a proper implementation should not necessary hold the start of the process for this. I think SDK can have a way to set this later.

**Comment by bogdandrutu** [comment]
@SergeyKanzhelev also that is one of the reason we want to default read the Resource from env variables to ensure we don't block the process to much.

**Comment by SergeyKanzhelev** [comment]
so resource auto-discovery modules would depend on SDK? I'm fine with it, just want to make sure it's understood.

Will there be any mention of Resource on API than? Only for `SpanData`?

**Comment by bogdandrutu** [comment]
I think we want people to add more labels, as discussed in https://github.com/open-telemetry/opentelemetry-specification/issues/10. These "resources" will be merged with the auto-detected by the SDK.

**Comment by SergeyKanzhelev** [comment]
@bogdandrutu than it probably should be `merge`, not `set`. Should we have neither `get` nor `set` for now and discuss what we want to do later as a feedback?

**Comment by pavolloffay** [comment]
done in https://github.com/open-telemetry/opentelemetry-java/pull/387

**Pull request closed without merging**